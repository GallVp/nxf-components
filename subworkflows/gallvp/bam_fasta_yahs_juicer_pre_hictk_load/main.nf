include { SAMTOOLS_FAIDX                    } from '../../../modules/gallvp/samtools/faidx/main'
include { JUICEBOXSCRIPTS_MAKEAGPFROMFASTA  } from '../../../modules/gallvp/juiceboxscripts/makeagpfromfasta/main'
include { YAHS_JUICERPRE                    } from '../../../modules/gallvp/yahs/juicerpre/main'
include { CUSTOM_YAHSJUICERPRE2PAIRS        } from '../../../modules/gallvp/custom/yahsjuicerpre2pairs/main'
include { HICTK_LOAD                        } from '../../../modules/gallvp/hictk/load/main'
include { HICTK_ZOOMIFY                     } from '../../../modules/gallvp/hictk/zoomify/main'
include { CUSTOM_YAHSJUICERPRE2TRACKS       } from '../../../modules/gallvp/custom/yahsjuicerpre2tracks/main'

workflow BAM_FASTA_YAHS_JUICER_PRE_HICTK_LOAD {

    take:
    ch_bam                                  // channel: [ val(meta), bam ]; meta: [ id, ref_id ]
    ch_fasta                                // channel: [ val(meta2), fasta ]; meta2: [ id ]
                                            // meta2.id = meta.ref_id
    val_assembly_mode                       // true|false

    main:
    ch_versions                             = Channel.empty()

    // MODULE: SAMTOOLS_FAIDX
    SAMTOOLS_FAIDX ( ch_fasta,
        [ [] , []],
        true // get_sizes
    )

    ch_fasta_fai                            = ch_fasta
                                            | join(SAMTOOLS_FAIDX.out.fai)
    ch_versions                             = ch_versions.mix(SAMTOOLS_FAIDX.out.versions.first())

    // MODULE: JUICEBOXSCRIPTS_MAKEAGPFROMFASTA
    JUICEBOXSCRIPTS_MAKEAGPFROMFASTA ( ch_fasta )

    ch_fasta_fai_agp                        = ch_fasta_fai
                                            | join(JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.agp)
    ch_versions                             = ch_versions.mix(JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.versions.first())

    // MODULE: YAHS_JUICERPRE
    ch_yahs_juicerpre_inputs                = ch_bam
                                            | map { meta, bam ->
                                                [ meta.ref_id, meta, bam ]
                                            }
                                            | join(
                                                ch_fasta_fai_agp
                                                | map { meta2, fasta, fai, agp ->
                                                    [ meta2.id, fasta, fai, agp ]
                                                }
                                            )
                                            | multiMap { _ref_id, meta, bam, _fasta, fai, agp ->
                                                bam: [ meta, bam ]
                                                agp: agp
                                                fai: fai
                                            }

    YAHS_JUICERPRE (
        ch_yahs_juicerpre_inputs.bam,
        ch_yahs_juicerpre_inputs.agp,
        ch_yahs_juicerpre_inputs.fai
    )

    ch_versions                             = ch_versions.mix(YAHS_JUICERPRE.out.versions.first())

    // MODULE: CUSTOM_YAHSJUICERPRE2PAIRS
    ch_yahsjuicerpre2pairs_in               = val_assembly_mode
                                            ? YAHS_JUICERPRE.out.txt
                                            | join(
                                                YAHS_JUICERPRE.out.sizes
                                            )
                                            | multiMap { meta, txt, sizes ->
                                                txt: [ meta, txt ]
                                                sizes: sizes
                                            }
                                            : YAHS_JUICERPRE.out.txt
                                            | map { meta, txt ->
                                                [ meta.ref_id, meta, txt ]
                                            }
                                            | join(
                                                SAMTOOLS_FAIDX.out.sizes.map { meta2, sizes -> [ meta2.id, sizes ] }
                                            )
                                            | multiMap { _ref_id, meta, txt, sizes ->
                                                txt: [ meta, txt ]
                                                sizes: sizes
                                            }

    CUSTOM_YAHSJUICERPRE2PAIRS (
        ch_yahsjuicerpre2pairs_in.txt,
        ch_yahsjuicerpre2pairs_in.sizes
    )

    ch_versions                             = ch_versions.mix(CUSTOM_YAHSJUICERPRE2PAIRS.out.versions.first())

    // MODULE: HICTK_LOAD
    HICTK_LOAD (
        CUSTOM_YAHSJUICERPRE2PAIRS.out.pairs,
        '4dn' // format
    )

    ch_versions                             = ch_versions.mix(HICTK_LOAD.out.versions.first())

    // MODULE: HICTK_ZOOMIFY
    HICTK_ZOOMIFY (
        HICTK_LOAD.out.hic
    )

    ch_versions                             = ch_versions.mix(HICTK_ZOOMIFY.out.versions.first())

    // MODULES: CUSTOM_ASSEMBLY2BEDPE
    ch_tracks_input                         = YAHS_JUICERPRE.out.liftover_agp
                                            | join(
                                                YAHS_JUICERPRE.out.scale
                                            )
                                            | multiMap { meta, agp, scale ->
                                                agp: [ meta, agp ]
                                                scale: scale
                                            }
    CUSTOM_YAHSJUICERPRE2TRACKS (
        ch_tracks_input.agp,
        ch_tracks_input.scale
    )

    ch_versions                             = ch_versions
                                            | mix(CUSTOM_YAHSJUICERPRE2TRACKS.out.versions.first())

    emit:
    hic                                     = HICTK_ZOOMIFY.out.hic                     // channel: [ meta, hic ]
    scale                                   = YAHS_JUICERPRE.out.scale                  // channel: [ meta, scale ]
    assembly                                = CUSTOM_YAHSJUICERPRE2TRACKS.out.assembly  // channel: [ meta, assembly ]
    bedpe                                   = CUSTOM_YAHSJUICERPRE2TRACKS.out.bedpe     // channel: [ meta, bedpe ]
    bed                                     = CUSTOM_YAHSJUICERPRE2TRACKS.out.bed       // channel: [ meta, bed ]
    versions                                = ch_versions                               // channel: [ versions.yml ]
}
