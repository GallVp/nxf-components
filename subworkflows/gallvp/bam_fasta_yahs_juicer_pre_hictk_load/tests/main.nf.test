nextflow_workflow {

    name "Test Subworkflow BAM_FASTA_YAHS_JUICER_PRE_HICTK_LOAD"
    script "../main.nf"
    config "./nextflow.config"
    workflow "BAM_FASTA_YAHS_JUICER_PRE_HICTK_LOAD"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/bam_fasta_yahs_juicer_pre_hictk_load"
    tag "samtools"
    tag "samtools/faidx"
    tag "yahs/juicerpre"
    tag "gunzip"
    tag "juiceboxscripts/makeagpfromfasta"
    tag "custom/yahsjuicerpre2tracks"
    tag "custom/yahsjuicerpre2pairs"
    tag "hictk/load"
    tag "hictk/zoomify"

    setup {
        run("GUNZIP") {
            script "../../../../modules/gallvp/gunzip/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [ id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.fasta.gz', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("hongyang - bam - fasta") {

        when {
            params {
                juicerpre = ''
                hictk_load = '--transpose-lower-triangular-pixels --bin-size 1kbp'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', ref_id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.bam', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = false // val_assembly_mode
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("hongyang - bam - fasta - assembly") {

        when {
            params {
                juicerpre = '-a'
                hictk_load = '--transpose-lower-triangular-pixels --bin-size 1kbp'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', ref_id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.bam', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = true // val_assembly_mode
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
