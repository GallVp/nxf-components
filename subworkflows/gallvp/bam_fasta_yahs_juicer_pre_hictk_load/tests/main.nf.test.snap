{
    "hongyang - bam - fasta - assembly": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.zoomify.hic:md5,1f86e06d96dcc9447c46f24d26a75240"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.scale:md5,b026324c6904b2a9cb4b88d6d61c81d1"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.assembly:md5,72dfdabd22760a91b81ed4241807a0e7"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.bedpe:md5,a4ec178eaaa2e96b724f927ce9678b5a"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.bed:md5,bc903fc7f717a627853e93ff6c2dfcb3"
                    ]
                ],
                "5": [
                    "versions.yml:md5,30dca07135ea1be722982fd223de2599",
                    "versions.yml:md5,4f0d0893fdc40f473fbf9dd1a6581a97",
                    "versions.yml:md5,7a2f7ac3fbf153b5b092f109093d568b",
                    "versions.yml:md5,843bdb19bb0e0f7ab0e9445463162d9b",
                    "versions.yml:md5,8cf024f69f36fa4269d30321e9da5be6",
                    "versions.yml:md5,d77c2f45baca90735e7eae7c3ee5bfe6",
                    "versions.yml:md5,ed972988c07090a7497d4d154e833fd0"
                ],
                "assembly": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.assembly:md5,72dfdabd22760a91b81ed4241807a0e7"
                    ]
                ],
                "bed": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.bed:md5,bc903fc7f717a627853e93ff6c2dfcb3"
                    ]
                ],
                "bedpe": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.bedpe:md5,a4ec178eaaa2e96b724f927ce9678b5a"
                    ]
                ],
                "hic": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.zoomify.hic:md5,1f86e06d96dcc9447c46f24d26a75240"
                    ]
                ],
                "scale": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.scale:md5,b026324c6904b2a9cb4b88d6d61c81d1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,30dca07135ea1be722982fd223de2599",
                    "versions.yml:md5,4f0d0893fdc40f473fbf9dd1a6581a97",
                    "versions.yml:md5,7a2f7ac3fbf153b5b092f109093d568b",
                    "versions.yml:md5,843bdb19bb0e0f7ab0e9445463162d9b",
                    "versions.yml:md5,8cf024f69f36fa4269d30321e9da5be6",
                    "versions.yml:md5,d77c2f45baca90735e7eae7c3ee5bfe6",
                    "versions.yml:md5,ed972988c07090a7497d4d154e833fd0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-21T08:57:25.842189"
    },
    "hongyang - bam - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.zoomify.hic:md5,fdf3100373d6683fcdd2ea95a9ba18cc"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    "versions.yml:md5,30dca07135ea1be722982fd223de2599",
                    "versions.yml:md5,7a2f7ac3fbf153b5b092f109093d568b",
                    "versions.yml:md5,843bdb19bb0e0f7ab0e9445463162d9b",
                    "versions.yml:md5,8cf024f69f36fa4269d30321e9da5be6",
                    "versions.yml:md5,d77c2f45baca90735e7eae7c3ee5bfe6",
                    "versions.yml:md5,ed972988c07090a7497d4d154e833fd0"
                ],
                "assembly": [
                    
                ],
                "bed": [
                    
                ],
                "bedpe": [
                    
                ],
                "hic": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.zoomify.hic:md5,fdf3100373d6683fcdd2ea95a9ba18cc"
                    ]
                ],
                "scale": [
                    
                ],
                "versions": [
                    "versions.yml:md5,30dca07135ea1be722982fd223de2599",
                    "versions.yml:md5,7a2f7ac3fbf153b5b092f109093d568b",
                    "versions.yml:md5,843bdb19bb0e0f7ab0e9445463162d9b",
                    "versions.yml:md5,8cf024f69f36fa4269d30321e9da5be6",
                    "versions.yml:md5,d77c2f45baca90735e7eae7c3ee5bfe6",
                    "versions.yml:md5,ed972988c07090a7497d4d154e833fd0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-21T08:57:12.171627"
    }
}